rfc/custom-metric-modules.md

```diff
@@ -0,0 +1,196 @@
+
+```go
+type MetricInfo struct {
+    // core.OperationContext contains all information needed including schema usage info
+	Operation core.OperationContext // making core.operationContext accessible outside of its package
+}
+
+type MetricRouterPreRequestHook interface {
+	EnterRouterRequest(ctx context.Context, info MetricInfo) error
+}
+
+type MetricRouterPostRequestHook interface {
+	LeaveRouterRequest(ctx context.Context, err error, statusCode int, info MetricInfo) error
+}
+
+type MetricRouterRequestHook interface {
+	MetricRouterPreRequestHook
+	MetricRouterPostRequestHook
+}
+
+type MetricSubgraphPreRequestHook interface {
+	EnterSubgraphRequest(ctx context.Context, info MetricInfo) error
+}
+
+type MetricSubgraphPostRequestHook interface {
+	LeaveSubgraphRequest(ctx context.Context, err error, statusCode int, info MetricInfo) error
+}
+
+type MetricSubgraphRequestHook interface {
+	MetricSubgraphPreRequestHook
+	MetricSubgraphPostRequestHook
+}
+
+type MetricOperationPreParseHook interface {
+	EnterOperationParse(ctx context.Context, info MetricInfo) error
+}
+
+type MetricOperationPostParseHook interface {
+	LeaveOperationParse(ctx context.Context, err error, statusCode int, info MetricInfo) error
+}
+
+type MetricOperationParseHook interface {
+	MetricOperationPreParseHook
+	MetricOperationPostParseHook
+}
+
+type MetricOperationPreNormalizeHook interface {
+	EnterOperationNormalize(ctx context.Context, info MetricInfo) error
+}
+
+type MetricOperationPostNormalizeHook interface {
+	LeaveOperationNormalize(ctx context.Context, err error, statusCode int, info MetricInfo) error
+}
+
+type MetricOperationNormalizeHook interface {
+	MetricOperationPreNormalizeHook
+	MetricOperationPostNormalizeHook
+}
+
+type MetricOperationPreValidateHook interface {
+	EnterOperationValidate(ctx context.Context, info MetricInfo) error
+}
+
+type MetricOperationPostValidateHook interface {
+	LeaveOperationValidate(ctx context.Context, err error, statusCode int, info MetricInfo) error
+}
+
+type MetricOperationValidateHook interface {
+	MetricOperationPreValidateHook
+	MetricOperationPostValidateHook
+}
+
+type MetricOperationPrePlanHook interface {
+	EnterOperationPlan(ctx context.Context, info MetricInfo) error
+}
+
+type MetricOperationPostPlanHook interface {
+	LeaveOperationPlan(ctx context.Context, err error, statusCode int, info MetricInfo) error
+}
+
+type MetricOperationPlanHook interface {
+	MetricOperationPrePlanHook
+	MetricOperationPostPlanHook
+}
+
+type MetricOperationPreExecuteHook interface {
+	EnterOperationExecute(ctx context.Context, info MetricInfo) error
+}
+
+type MetricOperationPostExecuteHook interface {
+	LeaveOperationExecute(ctx context.Context, err error, statusCode int, info MetricInfo) error
+}
+
+type MetricOperationExecuteHook interface {
+	MetricOperationPreExecuteHook
+	MetricOperationPostExecuteHook
+}
+
+type MetricOperationHook interface {
```

Comment:
What was your motivation to create a different abstraction of hooks just for metrics? Wouldn't it be easier and more consistent, if a developer can just leverage the standard hooks to implement custom metrics?
In my opinion, creating a second hook system just for metrics isn't needed.